src/thor/expansion_action.cc

```diff
@@ -78,43 +118,56 @@ std::string thor_worker_t::expansion(Api& request) {
   // a lambda that the path algorithm can call to add stuff to the dom
   // route and isochrone produce different GeoJSON properties
   std::string algo = "";
-  auto track_expansion = [&expansion, &opp_edges, &gen_factor, &skip_opps, &exp_props,
-                          &algo](baldr::GraphReader& reader, baldr::GraphId edgeid,
-                                 baldr::GraphId prev_edgeid, const char* algorithm = nullptr,
-                                 const Expansion_EdgeStatus status = Expansion_EdgeStatus_reached,
-                                 const float duration = 0.f, const uint32_t distance = 0,
-                                 const float cost = 0.f) {
-    algo = algorithm;
-
-    auto tile = reader.GetGraphTile(edgeid);
-    if (tile == nullptr) {
-      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
-                std::to_string(edgeid.Tile_Base()));
-      return;
-    }
-    const auto* edge = tile->directededge(edgeid);
-    // unfortunately we have to call this before checking if we can skip
-    // else the tile could change underneath us when we get the opposing
-    auto shape = tile->edgeinfo(edge).shape();
-    auto names = tile->edgeinfo(edge).GetNames();
-    auto is_forward = edge->forward();
-
-    // if requested, skip this edge in case its opposite edge has been added
-    // before (i.e. lower cost) else add this edge's id to the lookup container
-    if (skip_opps) {
-      auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, tile);
-      if (opp_edgeid && opp_edges.count(opp_edgeid))
-        return;
-      opp_edges.insert(edgeid);
-    }
-
-    if (!edge->forward())
-      std::reverse(shape.begin(), shape.end());
-    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
-
-    writeExpansionProgress(expansion, edgeid, prev_edgeid, shape, exp_props, status, duration,
-                           distance, cost);
-  };
+  auto track_expansion =
+      [&expansion, &opp_edges, &gen_factor, &skip_opps, &exp_props, &algo, &edge_state,
+       &dedupe](baldr::GraphReader& reader, baldr::GraphId edgeid, baldr::GraphId prev_edgeid,
+                const char* algorithm = nullptr,
+                const Expansion_EdgeStatus status = Expansion_EdgeStatus_reached,
+                const float duration = 0.f, const uint32_t distance = 0, const float cost = 0.f) {
+        algo = algorithm;
+
+        auto tile = reader.GetGraphTile(edgeid);
+        if (tile == nullptr) {
+          LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                    std::to_string(edgeid.Tile_Base()));
+          return;
+        }
```

Comment:
i know this was here before but it seems crazy pedantic. the graph reader can fail to get a tile for sure but not one its already seen (ie we know is in the tileset). the worst case is that its a cache miss and it has to pull it from disk or from network but if the algorithm is passing you this graph id for an edge its seen it will have to be able to get the edge. imho we can delete this check, if this happens someone is deleting tiles from the data storage while the application is running which is not worth protecting against

Reply:
fair